rust-src/id/src/id_proof_types.rs

```diff
@@ -0,0 +1,380 @@
+        )]
+        proof: SetMembershipProof<C>,
+    },
+    AttributeNotInSet {
+        #[serde(
+            rename = "proof",
+            serialize_with = "base16_encode",
+            deserialize_with = "base16_decode"
+        )]
+        proof: SetNonMembershipProof<C>,
+    },
+}
+
+/// A statement about a credential on an account.
+#[derive(Debug, Clone, SerdeSerialize, SerdeDeserialize)]
+#[serde(bound(
+    serialize = "C: Curve, AttributeType: Attribute<C::Scalar> + SerdeSerialize",
+    deserialize = "C: Curve, AttributeType: Attribute<C::Scalar> + SerdeDeserialize<'de>"
+))]
+pub struct StatementWithContext<C: Curve, AttributeType: Attribute<C::Scalar>> {
+    pub account:    AccountAddress,
+    #[serde(serialize_with = "base16_encode", deserialize_with = "base16_decode")]
+    pub credential: CredId<C>,
+    pub statement:  Statement<C, AttributeType>,
+}
+
+/// A secret needed for proving a statement
+#[derive(Debug, Clone, SerdeSerialize, SerdeDeserialize)]
+#[serde(bound(
+    serialize = "C: Curve, AttributeType: Attribute<C::Scalar> + SerdeSerialize",
+    deserialize = "C: Curve, AttributeType: Attribute<C::Scalar> + SerdeDeserialize<'de>"
+))]
+pub struct Secret<C: Curve, AttributeType: Attribute<C::Scalar>> {
+    pub secrets: Vec<AtomicSecret<C, AttributeType>>,
+}
+
+/// A statement is a list of atomic statements.
+#[derive(Debug, Clone, SerdeSerialize, SerdeDeserialize)]
+#[serde(bound(
+    serialize = "C: Curve, AttributeType: Attribute<C::Scalar> + SerdeSerialize",
+    deserialize = "C: Curve, AttributeType: Attribute<C::Scalar> + SerdeDeserialize<'de>"
+))]
+pub struct Statement<C: Curve, AttributeType: Attribute<C::Scalar>> {
+    pub statements: Vec<AtomicStatement<C, AttributeType>>,
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> Default for Statement<C, AttributeType> {
+    fn default() -> Self { Statement { statements: vec![] } }
+}
+
+/// Helper functions for constructing statements
+impl Statement<G1, AttributeKind> {
+    pub fn older_than(mut self, age: u64) -> Self {
+        use chrono::Datelike;
+        let now = chrono::Utc::now();
+        let date_years_ago = format!(
+            "{:04}{:02}{:02}",
+            (now.year() as u64) - age,
+            now.month(),
+            now.day()
+        );
+        let upper = AttributeKind(date_years_ago);
+        let lower = AttributeKind(String::from("18000101"));
+
+        let statement = AttributeInRangeStatement::<G1, _> {
+            attribute_tag: AttributeTag(3u8), // DOB tag
+            lower,
+            upper,
+            _phantom: PhantomData::default(),
+        };
+        self.statements
+            .push(AtomicStatement::AttributeInRange { statement });
+        self
+    }
+
+    pub fn younger_than(mut self, age: u64) -> Self {
+        use chrono::Datelike;
+        let now = chrono::Utc::now();
+        let date_years_ago = format!(
+            "{:04}{:02}{:02}",
+            (now.year() as u64) - age,
```

Comment:
Please do this more robustly, and return Option<Self> instead. And check for underflow.